### PR TITLE
feat: simd double unary math ops

### PR DESCRIPTION
**Pull request by clintropolis**
### Description
Adds a bunch of simd optimizations of the remaining double unary operators. The operators do not have direct intrinsics but are backed by a vectorized math library that depends on platform and jdk build, so performance may vary.

Along the way I found sinh, cosh, and tanh were not wired up to SQL, so have added those 
Testing with a M1 mac I generally saw a pretty nice improvement:

```
SELECT SUM(LN(double1)) FROM expressions
SELECT SUM(LN(long1)) FROM expressions
SELECT SUM(EXP(double1)) FROM expressions
SELECT SUM(EXP(long1)) FROM expressions
SELECT SUM(LOG10(double1)) FROM expressions
SELECT SUM(LOG10(long1)) FROM expressions
SELECT SUM(CBRT(double1)) FROM expressions
SELECT SUM(CBRT(long1)) FROM expressions
SELECT SUM(SIN(double1)) FROM expressions
SELECT SUM(SIN(long1)) FROM expressions
SELECT SUM(COS(double1)) FROM expressions
SELECT SUM(COS(long1)) FROM expressions
SELECT SUM(TAN(double1)) FROM expressions
SELECT SUM(TAN(long1)) FROM expressions
SELECT SUM(ASIN(double1)) FROM expressions
SELECT SUM(ASIN(long1)) FROM expressions
SELECT SUM(ACOS(double1)) FROM expressions
SELECT SUM(ACOS(long1)) FROM expressions
SELECT SUM(ATAN(double1)) FROM expressions
SELECT SUM(ATAN(long1)) FROM expressions
SELECT SUM(SINH(double1)) FROM expressions
SELECT SUM(SINH(long1)) FROM expressions
SELECT SUM(COSH(double1)) FROM expressions
SELECT SUM(COSH(long1)) FROM expressions
SELECT SUM(TANH(double1)) FROM expressions
SELECT SUM(TANH(long1)) FROM expressions
```

```
Benchmark                        (complexCompression)  (deferExpressionDimensions)  (jsonObjectStorageEncoding)  (query)  (rowsPerSegment)  (schemaType)  (storageType)  (stringEncoding)  (useVectorApi)  (vectorize)  Mode  Cnt   Score   Error  Units
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       71           1500000      explicit           MMAP              UTF8           false        force  avgt    5  30.285 ± 1.411  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       71           1500000      explicit           MMAP              UTF8            true        force  avgt    5  12.606 ± 2.843  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       72           1500000      explicit           MMAP              UTF8           false        force  avgt    5  28.189 ± 1.549  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       72           1500000      explicit           MMAP              UTF8            true        force  avgt    5  14.588 ± 0.554  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       73           1500000      explicit           MMAP              UTF8           false        force  avgt    5  29.777 ± 1.107  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       73           1500000      explicit           MMAP              UTF8            true        force  avgt    5  10.653 ± 0.623  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       74           1500000      explicit           MMAP              UTF8           false        force  avgt    5  22.288 ± 4.846  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       74           1500000      explicit           MMAP              UTF8            true        force  avgt    5  11.983 ± 0.744  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       75           1500000      explicit           MMAP              UTF8           false        force  avgt    5  35.076 ± 1.976  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       75           1500000      explicit           MMAP              UTF8            true        force  avgt    5  13.469 ± 0.812  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       76           1500000      explicit           MMAP              UTF8           false        force  avgt    5  33.361 ± 0.889  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       76           1500000      explicit           MMAP              UTF8            true        force  avgt    5  14.979 ± 0.777  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       77           1500000      explicit           MMAP              UTF8           false        force  avgt    5  21.475 ± 1.624  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       77           1500000      explicit           MMAP              UTF8            true        force  avgt    5  18.852 ± 0.987  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       78           1500000      explicit           MMAP              UTF8           false        force  avgt    5  23.074 ± 0.616  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       78           1500000      explicit           MMAP              UTF8            true        force  avgt    5  20.570 ± 1.671  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       79           1500000      explicit           MMAP              UTF8           false        force  avgt    5  33.913 ± 1.415  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       79           1500000      explicit           MMAP              UTF8            true        force  avgt    5  14.416 ± 1.236  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       80           1500000      explicit           MMAP              UTF8           false        force  avgt    5  28.261 ± 1.315  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       80           1500000      explicit           MMAP              UTF8            true        force  avgt    5  17.651 ± 1.072  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       81           1500000      explicit           MMAP              UTF8           false        force  avgt    5  34.117 ± 1.413  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       81           1500000      explicit           MMAP              UTF8            true        force  avgt    5  15.238 ± 0.762  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       82           1500000      explicit           MMAP              UTF8           false        force  avgt    5  28.001 ± 1.855  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       82           1500000      explicit           MMAP              UTF8            true        force  avgt    5  18.244 ± 0.971  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       83           1500000      explicit           MMAP              UTF8           false        force  avgt    5  46.805 ± 2.455  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       83           1500000      explicit           MMAP              UTF8            true        force  avgt    5  17.212 ± 0.605  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       84           1500000      explicit           MMAP              UTF8           false        force  avgt    5  47.951 ± 1.824  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       84           1500000      explicit           MMAP              UTF8            true        force  avgt    5  23.149 ± 0.962  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       85           1500000      explicit           MMAP              UTF8           false        force  avgt    5  28.384 ± 0.810  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       85           1500000      explicit           MMAP              UTF8            true        force  avgt    5  11.907 ± 1.053  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       86           1500000      explicit           MMAP              UTF8           false        force  avgt    5  22.639 ± 1.056  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       86           1500000      explicit           MMAP              UTF8            true        force  avgt    5  13.509 ± 0.615  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       87           1500000      explicit           MMAP              UTF8           false        force  avgt    5  29.227 ± 5.817  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       87           1500000      explicit           MMAP              UTF8            true        force  avgt    5  12.091 ± 0.466  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       88           1500000      explicit           MMAP              UTF8           false        force  avgt    5  22.917 ± 0.659  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       88           1500000      explicit           MMAP              UTF8            true        force  avgt    5  14.683 ± 1.046  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       89           1500000      explicit           MMAP              UTF8           false        force  avgt    5  34.433 ± 2.669  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       89           1500000      explicit           MMAP              UTF8            true        force  avgt    5  15.296 ± 0.497  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       90           1500000      explicit           MMAP              UTF8           false        force  avgt    5  27.284 ± 0.832  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       90           1500000      explicit           MMAP              UTF8            true        force  avgt    5  16.977 ± 1.177  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       91           1500000      explicit           MMAP              UTF8           false        force  avgt    5  38.460 ± 1.404  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       91           1500000      explicit           MMAP              UTF8            true        force  avgt    5  16.669 ± 0.656  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       92           1500000      explicit           MMAP              UTF8           false        force  avgt    5  22.888 ± 0.748  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       92           1500000      explicit           MMAP              UTF8            true        force  avgt    5  17.944 ± 1.868  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       93           1500000      explicit           MMAP              UTF8           false        force  avgt    5  35.114 ± 2.037  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       93           1500000      explicit           MMAP              UTF8            true        force  avgt    5  16.522 ± 1.252  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       94           1500000      explicit           MMAP              UTF8           false        force  avgt    5  22.599 ± 0.924  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       94           1500000      explicit           MMAP              UTF8            true        force  avgt    5  17.941 ± 0.951  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       95           1500000      explicit           MMAP              UTF8           false        force  avgt    5  35.517 ± 0.967  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       95           1500000      explicit           MMAP              UTF8            true        force  avgt    5  37.003 ± 1.007  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       96           1500000      explicit           MMAP              UTF8           false        force  avgt    5  20.983 ± 0.902  ms/op
SqlExpressionBenchmark.querySql                  NONE                 singleString                        SMILE       96           1500000      explicit           MMAP              UTF8            true        force  avgt    5  16.690 ± 0.919  ms/op
```

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/math/expr/vector/simd/SimdDoubleSinProcessor.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  }
+
+  @Override
+  protected void processVector(double[] input, boolean[] inputNulls, int currentSize)
+  {
+    final int laneCount = SPECIES.length();
+    final int upperBound = SPECIES.loopBound(currentSize);
+    int i = 0;
+    for (; i < upperBound; i += laneCount) {
+      DoubleVector.fromArray(SPECIES, input, i).lanewise(VectorOperators.SIN).intoArray(outValues, i);
```
Comment (FrankChen021) [inline processing/src/main/java/org/apache/druid/math/expr/vector/simd/SimdDoubleSinProcessor.java:48]: [P1] Keep transcendental results stable across JIT warmup

`VectorOperators.SIN` initially matches `Math.sin`, but compiled SLEEF/SVML output can differ. On JDK 25/AArch64, this exact loop changed 665 of 8,192 lanes after warmup; most other added `VO_MATHLIB` operations also diverged. Compilation mid-query can therefore give identical inputs different bits, affecting exact filters or group keys. Keep these operations scalar unless stable semantics are established, and add warmed parity tests.

**Review by FrankChen021 — COMMENTED** [review]
| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 1 |
| P2 | 0 |
| P3 | 0 |
| Total | 1 |

| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 1 |
| P2 | 0 |
| P3 | 0 |
| Total | 1 |

Reviewed 36 of 36 changed files.

---

This is an automated review by Codex GPT-5.6-Sol

File: processing/src/main/java/org/apache/druid/math/expr/vector/simd/SimdDoubleSinProcessor.java (hunk 1 above)
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/math/expr/vector/simd/SimdDoubleSinProcessor.java:48]: hmm, fair. It does seem pretty unlikely that someone is going to be using `sin(x)` for grouping or equality filtering, so as a compromise I've split these functions out into a separate config `useVectorMathApi` (which also requires `useVectorApi`) and documented it to call out this behavior is possible. These configs are both off by default, and this at least makes them available to operators who know they are safe since I am unsure how to reliably pre-warm compilation like this reliably at this time. Added a test that shows that after-warm-up eventually stabilizes.

**Review by FrankChen021 — COMMENTED** [review]
Follow-up complete: the stability concern is addressed. VO_MATHLIB operations now require the separate, default-off `useVectorMathApi` gate in addition to `useVectorApi`, and the documentation explicitly warns against exact-equality use cases. The warmed parity test also captures the expected stabilization behavior.

Reviewed 41 of 41 changed files.

Validation: `git diff --check` passed.

---

This is an automated review by Codex GPT-5.6-Sol

**Review by FrankChen021 — COMMENTED** [review]
I have reviewed the updated code for correctness, edge cases, concurrency, and integration risks; no issues found.

Reviewed 42 of 42 changed files. The default-off vector-math gate remains intact; the latest update only adds required spelling exceptions.

---

This is an automated review by Codex GPT-5.6-Sol

File: processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessing.java [hunk 2]
```diff
@@ -45,19 +45,19 @@ public class ExpressionProcessing
...
   {
-    INSTANCE = new ExpressionProcessingConfig(null, true, null, null);
+    INSTANCE = new ExpressionProcessingConfig(null, true, null, null, null);
   }
 
   @VisibleForTesting
   public static void initializeForVectorApiTests()
   {
-    INSTANCE = new ExpressionProcessingConfig(null, null, null, true);
+    INSTANCE = new ExpressionProcessingConfig(null, null, null, true, true);
```
Comment (FrankChen021) [inline processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessing.java:60]: [P1] Keep exact vector-math tests from enabling approximate SIMD

`initializeForVectorApiTests()` enables approximate math SIMD, but the inherited exact-comparison suite uses `Assert.assertEquals` for functions such as `sin` and `exp`. This PR documents VO_MATHLIB ULP differences and tier-transition changes, so supported JDK/hardware combinations can now fail these tests. Keep this helper's math flag off, or make the consistency assertions tolerance-aware.

**Review by FrankChen021 — COMMENTED** [review]
| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 1 |
| P2 | 0 |
| P3 | 0 |
| Total | 1 |

| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 1 |
| P2 | 0 |
| P3 | 0 |
| Total | 1 |

Reviewed 42 of 42 changed files. The prior default-off stability finding is resolved; this updated review reports one new test-gating regression.



---

This is an automated review by Codex GPT-5.6-Luna(max)

File: processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessingConfig.java [hunk 3]
```diff
@@ -70,6 +75,7 @@ public ExpressionProcessingConfig(
         "true"
     );
     this.useVectorApi = getWithPropertyFallbackFalse(useVectorApi, USE_VECTOR_API);
+    this.useVectorMathApi = getWithPropertyFallbackFalse(useVectorMathApi, USE_VECTOR_MATH_API);
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessingConfig.java]: IMO this should be `true` by default, since I believe most people that enable `useVectorApi` will want this. Setting it to `false` is a niche thing.

File: processing/src/main/java/org/apache/druid/math/expr/vector/simd/SimdDoubleAcosProcessor.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import org.apache.druid.math.expr.vector.ExprVectorProcessor;
+import org.apache.druid.math.expr.vector.functional.DoubleUnivariateDoubleFunction;
+
+import java.util.Arrays;
+
+/**
+ * SIMD specialization of {@code (double[]) -> double[]} arc cosine. See {@link SimdSupportedUnaryOp#ACOS} for
+ * the shared VO_MATHLIB performance notes.
+ */
+public final class SimdDoubleAcosProcessor extends SimdDoubleUnaryProcessor
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/math/expr/vector/simd/SimdDoubleAcosProcessor.java:34]: There are so many of these. I wonder if generating the classes at runtime would be better than having all the sources here. Could be something to explore in a follow-up.

File: processing/src/main/java/org/apache/druid/math/expr/vector/SimpleVectorMathUnivariateDoubleProcessorFactory.java [hunk 5]
```diff
@@ -64,10 +66,20 @@ protected SimpleVectorMathUnivariateDoubleProcessorFactory(
...
   }
 
+  private boolean simdEnabled()
+  {
+    if (simdOp == null || !ExpressionProcessing.useVectorApi()) {
+      return false;
+    }
+    // VO_MATHLIB ops (SVML/SLEEF-backed transcendentals) sit behind the additional useVectorMathApi opt-in
+    // because they can produce different bits than Math.<op> once JIT tier transitions the loop from C1 to C2.
+    return !simdOp.isMathLib() || ExpressionProcessing.useVectorMathApi();
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/math/expr/vector/SimpleVectorMathUnivariateDoubleProcessorFactory.java]: Should probably have a version of this check in `SimdSupportedBinaryOp` and `SimpleVectorMathBivariateProcessorFactory` too. Maybe nothing would return true today, but it seems useful to have it there so it's not forgotten in the future when more ops are added.

Or, possibly even better, move this all into `simdOp.isSimdEnabled()`.

File: processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessing.java [hunk 6]
```diff
@@ -99,6 +99,22 @@ public static boolean useVectorApi()
     return INSTANCE.useVectorApi();
   }
 
+  /**
+   * Whether SIMD dispatch is allowed for math ops backed by the JDK's VO_MATHLIB path (LOG, EXP, SIN, etc). Off by
+   * default; opt-in via {@link ExpressionProcessingConfig#USE_VECTOR_MATH_API}, and additionally requires
+   * {@link #useVectorApi()}.
+   *
+   * <p>These ops route through Intel SVML / Arm SLEEF once the JIT compiles the vector loop to C2; before that
+   * compilation, they fall back to per-lane {@link Math} calls. The two paths can differ by up to 1 ulp, so a
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessing.java]: This says up to 1 ulp, but `configuration/index.md` says "a few ulps" and the test `SimdVoMathlibParityTest.MAX_ULPS` is checking for 4 ulps. Please make all the commentary about accuracy consistent.

File: processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessing.java [hunk 7]
```diff
@@ -99,6 +99,22 @@ public static boolean useVectorApi()
     return INSTANCE.useVectorApi();
   }
 
+  /**
+   * Whether SIMD dispatch is allowed for math ops backed by the JDK's VO_MATHLIB path (LOG, EXP, SIN, etc). Off by
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessing.java]: IMO should be on by default (see other comment).

File: docs/configuration/index.md [hunk 8]
```diff
@@ -652,6 +652,7 @@ the `expression` aggregator/post-aggregator, and any SQL functions that lower to
...
 |`druid.expressions.useVectorApi`|If true, vectorized expression vector processors and numeric vector aggregators dispatch to SIMD specializations backed by the JDK incubator Vector API (`jdk.incubator.vector`) where available. Requires `--add-modules=jdk.incubator.vector` on the JVM command line (see [strong encapsulation](../operations/java.md#strong-encapsulation)). Off by default while the Vector API remains an incubator JDK feature.|false|
+|`druid.expressions.useVectorMathApi`|If true (and `useVectorApi` is also true), math expressions whose SIMD path is backed by the JDK's VO_MATHLIB routing (Intel SVML on x86, SLEEF on Arm) dispatch to that SIMD path. Results can differ from the scalar `Math.<op>` by up to a few ulps, and the exact bits can shift once the JIT promotes the SIMD loop from C1 to C2, meaning a long-running query can produce different bits for the same input across the tier transition. Safe for `SUM`/`MIN`/`MAX`/range-filter workloads; **do not enable** if queries rely on exact-bit equality of these functions (for example `GROUP BY sin(x)` where all rows with the same `x` must land in the same group, or `WHERE sin(x) = sin(y)`). Off by default.|false|
```
Comment (gianm) [inline docs/configuration/index.md]: We definitely shouldn't encourage users to rely on exact-bit equality for floating point numbers. It's better to suggest that people truncate or round for comparison.

File: processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessing.java (hunk 2 above)
Comment (gianm) [reply processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessing.java:60]: This seems like it could be a legit issue with `VectorExprResultConsistencyVectorApiTest`, possibly dependent on the order in which things are run.

File: processing/src/main/java/org/apache/druid/math/expr/vector/simd/SimdDoubleAcosProcessor.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+  protected void processVector(double[] input, boolean[] inputNulls, int currentSize)
+  {
+    final int laneCount = SPECIES.length();
+    final int upperBound = SPECIES.loopBound(currentSize);
+    int i = 0;
+    for (; i < upperBound; i += laneCount) {
+      DoubleVector.fromArray(SPECIES, input, i).lanewise(VectorOperators.ACOS).intoArray(outValues, i);
+    }
+    for (; i < currentSize; i++) {
+      outValues[i] = scalarFallback.process(input[i]);
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/math/expr/vector/simd/SimdDoubleAcosProcessor.java]: This approach of doing scalar fallbacks means that the accuracy discrepancies are more pervasive than the docs in `configuration/index.md` suggest. The final vector in a cursor (which is likely to be shorter than the max size) will be using different math ops than the other vectors. Probably the best way to fix it is to use a vector op here too, with a mask.

File: processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessing.java (hunk 2 above)
Comment (FrankChen021) [reply processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessing.java:60]: The issue remains at current HEAD: `initializeForVectorApiTests()` still enables `useVectorMathApi` at line 60, while the inherited consistency suite compares SIMD and scalar math with exact equality. Please disable this flag in that helper or make the affected assertions tolerance-aware.

Reviewed 42 of 42 changed files.

File: processing/src/test/java/org/apache/druid/math/expr/vector/VectorTestAssertions.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  {
+    if (Double.isNaN(expected) && Double.isNaN(actual)) {
+      return;
+    }
+    if (Double.doubleToRawLongBits(expected) == Double.doubleToRawLongBits(actual)) {
+      return;
+    }
+    final double ulp = Math.ulp(expected);
+    final double diff = Math.abs(actual - expected);
+    if (Double.isNaN(diff) || diff > (double) maxUlps * ulp) {
```
Comment (FrankChen021) [inline processing/src/test/java/org/apache/druid/math/expr/vector/VectorTestAssertions.java:65]: P2 Handle non-finite expected values before ULP comparison

When expected is infinite and actual is finite or has the opposite sign, Math.ulp(expected) and the comparison threshold are infinite, so this assertion can incorrectly pass. Handle non-finite values with exact equality before calculating the ULP threshold so parity tests fail when an infinite expected result is not reproduced.

**Review by FrankChen021 — COMMENTED** [review]
| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 0 |
| P2 | 1 |
| P3 | 0 |
| Total | 1 |

Reviewed 46 of 46 changed files. The prior test-gating concern appears resolved; this updated review found one test-helper edge case.

Validation: focused `git diff --no-ext-diff --check` passed. Builds and tests were not run.



---

This is an automated review by Codex GPT-5.6-Luna(max)

File: benchmarks/src/test/java/org/apache/druid/benchmark/query/SqlExpressionBenchmark.java [hunk 11]
```diff
@@ -179,7 +179,34 @@ public class SqlExpressionBenchmark extends SqlBaseQueryBenchmark
...
-      "SELECT SUM(SQRT(long1)) FROM expressions"
+      "SELECT SUM(SQRT(long1)) FROM expressions",
+      // 71-96: VO_MATHLIB unary ops exposed via SQL and covered by SIMD dispatch.
+      "SELECT SUM(LN(double1)) FROM expressions",
+      "SELECT SUM(LN(long1)) FROM expressions",
+      "SELECT SUM(EXP(double1)) FROM expressions",
+      "SELECT SUM(EXP(long1)) FROM expressions",
+      "SELECT SUM(LOG10(double1)) FROM expressions",
+      "SELECT SUM(LOG10(long1)) FROM expressions",
+      "SELECT SUM(CBRT(double1)) FROM expressions",
```
Comment (FrankChen021) [inline benchmarks/src/test/java/org/apache/druid/benchmark/query/SqlExpressionBenchmark.java:190]: [P2] CBRT benchmark queries are not plannable

The two new queries call CBRT, but CBRT is not registered in DruidOperatorTable. These benchmark cases therefore fail during SQL planning instead of exercising the intended SIMD expression path. Add the operator conversion/registration or remove the cases until the function is supported by the planner.

**Review by FrankChen021 — COMMENTED** [review]
| Severity | Findings |
| --- | ---: |
| P0 | 0 |
| P1 | 0 |
| P2 | 1 |
| P3 | 0 |
| Total | 1 |

The latest test-helper non-finite comparison issue is resolved, but the new CBRT benchmark cases are not plannable.

Reviewed 1 of 46 changed files with findings; all 46 changed files were reviewed.

Validation: focused `git diff --check` passed. Builds and tests were not run.



---

This is an automated review by Codex GPT-5.6-Luna(max)

File: benchmarks/src/test/java/org/apache/druid/benchmark/query/SqlExpressionBenchmark.java (hunk 11 above)
Comment (clintropolis) [reply benchmarks/src/test/java/org/apache/druid/benchmark/query/SqlExpressionBenchmark.java:190]: good catch, fixed

File: docs/configuration/index.md [hunk 12]
```diff
@@ -652,7 +652,7 @@ the `expression` aggregator/post-aggregator, and any SQL functions that lower to
...
+|`druid.expressions.useVectorMathApi`|When `useVectorApi` is true, controls whether math expressions whose SIMD path is backed by the JDK's VO_MATHLIB routing (Intel SVML on x86, SLEEF on Arm) dispatch to that SIMD path. Results can differ from the scalar `Math.<op>` by a few ulps (the parity test bounds this at 2 ulps), and the exact bits can shift once the JIT promotes the SIMD loop from C1 to C2 during a long-running query. In general applications should not rely on bit-for-bit equality of floating-point results, and round or truncate first when using for grouping or equality comparisons. This flag exists as an escape hatch for legacy queries that rely on `Math.<op>` bits without rounding. Has no effect unless `useVectorApi` is also true. On by default when `useVectorApi` is enabled.|true|
```
Comment (gianm) [inline docs/configuration/index.md]: The docs shouldn't mention "the parity test", that's a developer thing. The docs are for users. If we can't offer a specific bound here then try language like "SIMD and non-SIMD results can differ slightly"

File: processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessing.java (hunk 2 above)
Comment (gianm) [reply processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessing.java:60]: @clintropolis could you please opine on this?

**Review by FrankChen021 — COMMENTED** [review]
Follow-up review complete. The exact-equality concern is addressed at current HEAD: `VectorExprResultConsistencyTest` now compares `Double` results with a 2-ULP tolerance. The CBRT benchmark planning concern is also resolved by registering CBRT and adding planner coverage.

Reviewed 46 of 46 changed files.

Validation: focused `git diff --no-ext-diff --check` passed. Builds and tests were not run.



---

This is an automated review by Codex GPT-5.6-Luna(max)

**Review by FrankChen021 — COMMENTED** [review]
I have reviewed the code for correctness, edge cases, concurrency, and integration risks; no issues found.

Reviewed 46 of 46 changed files.

---

This is an automated review by Codex GPT-5.6-Luna(max)

File: processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessing.java (hunk 2 above)
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessing.java:60]: oh, i fixed the tests to do the same 2 ulp variance (tests which could be impacted use `VectorTestAssertions.assertDoublesEquivalent`)

Comment (FrankChen021) [reply processing/src/main/java/org/apache/druid/math/expr/ExpressionProcessing.java:60]: Thanks, this resolves the concern: the inherited vector-API consistency suite now compares scalar Double results with VectorTestAssertions.assertDoublesEquivalent(..., 2), so it no longer requires bit-for-bit equality for VO_MATHLIB operations. I reviewed all 46 changed files and found no additional PR-caused issues.

**Review by FrankChen021 — COMMENTED** [review]
The prior SIMD exact-comparison concern is resolved in the current head; the author updated the affected assertions to allow the documented 2-ULP variance. No new issues found.

Reviewed 46 of 46 changed files.



---

This is an automated review by Codex GPT-5.6-Luna(max)

File: processing/src/test/java/org/apache/druid/math/expr/vector/simd/SimdVoMathlibParityTest.java [hunk 13]
```diff
@@ -0,0 +1,417 @@
...
+  // we start capturing "final" output.
+  private static final int WARMUP_ITERATIONS = 30_000;
+  private static final int STABILITY_ITERATIONS = 500;
+  // Small integer multiple of any reasonable DoubleVector.SPECIES_PREFERRED length (2/4/8) so the SIMD
+  // loop takes multiple iterations per invocation.
+  private static final int INPUT_LANES = 64;
+  // 2 ulps is spec-defensible for the SIMD-vs-scalar comparison: Java's Math is spec'd to be within 1 ulp
+  // of the correctly-rounded result for most transcendentals (sin/cos/tan/log/exp/asin/acos/atan/cbrt/...);
+  // the SIMD path targets the same tolerance, so the SIMD-vs-scalar delta can be up to 2 ulps just from
+  // both landing at opposite ends of their independent 1-ulp windows. sinh/cosh/tanh have no strict Math
```
Comment (gianm) [inline processing/src/test/java/org/apache/druid/math/expr/vector/simd/SimdVoMathlibParityTest.java:60]: On my JDK at least, the javadoc for `Math.sinh`, `cosh`, and `tanh` specify that they are within 2.5 ulps of the correct result.

File: sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidOperatorTable.java [hunk 14]
```diff
@@ -391,6 +391,10 @@ public class DruidOperatorTable implements SqlOperatorTable
                    .add(new DirectOperatorConversion(SqlStdOperatorTable.ACOS, "acos"))
                    .add(new DirectOperatorConversion(SqlStdOperatorTable.ATAN, "atan"))
                    .add(new DirectOperatorConversion(SqlStdOperatorTable.ATAN2, "atan2"))
+                   .add(new DirectOperatorConversion(SqlLibraryOperators.SINH, "sinh"))
```
Comment (gianm) [inline sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidOperatorTable.java:394]: The new functions should be added to the docs.

File: processing/src/test/java/org/apache/druid/math/expr/VectorExprResultConsistencyTest.java [hunk 15]
```diff
@@ -828,11 +829,17 @@ public static void assertEvalsMatch(
...
           );
         } else {
-          Assertions.assertEquals(
-              nonVectorEval.valueOrThrow()[i],
-              vectorEval.valueOrThrow()[i],
-              message
-          );
+          final Object expected = nonVectorEval.valueOrThrow()[i];
+          final Object actual = vectorEval.valueOrThrow()[i];
+          // Double values compared with ulp-tolerance so the VO_MATHLIB SIMD path can differ from the
```
Comment (gianm) [inline processing/src/test/java/org/apache/druid/math/expr/VectorExprResultConsistencyTest.java:834]: Is it possible to do this loosening only for the mathlib stuff that requires it?

File: processing/src/main/java/org/apache/druid/math/expr/vector/simd/SimdSupportedUnaryOp.java [hunk 16]
```diff
@@ -19,18 +19,117 @@
...
+  public boolean isMathLib()
+  {
+    return mathLib;
+  }
+
+  /**
+   * Whether SIMD dispatch is currently enabled for this op according to the runtime
+   * {@link ExpressionProcessing#useVectorApi()} / {@link ExpressionProcessing#useVectorMathApi()} flags.
+   */
+  public boolean isSimdEnabled()
```
Comment (gianm) [inline processing/src/main/java/org/apache/druid/math/expr/vector/simd/SimdSupportedUnaryOp.java:128]: Looks like there's one more old call site to convert in `SimpleVectorMathUnivariateProcessorFactory`.
